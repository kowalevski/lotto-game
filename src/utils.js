/* eslint-disable no-plusplus */
import { v4 as uuid } from 'uuid';

const EMPTY_CELLS_MAX_COUNT = {
  ROW: 4,
  COLUMN: 2
};

const CELLS_ROW_MAX_COUNT = 9;
const ROWS_MAX_COUNT = 3;

const NUMBERS_RANGE = {
  MIN: 1,
  MAX: 90
};

const getRandomInt = (argMin, argMax, exclude) => {
  const min = Math.ceil(argMin);
  const max = Math.floor(argMax);
  const res = Math.floor(Math.random() * (max - min)) + min;

  return exclude.includes(res) ? getRandomInt(argMin, argMax, exclude) : res;
};

const isEmptyCell = (rowEmptyCells, columnEmptyCells) => {
  const isEmpty = Math.random() >= 0.5;

  return (
    (!rowEmptyCells || rowEmptyCells < EMPTY_CELLS_MAX_COUNT.ROW) &&
    (!columnEmptyCells || columnEmptyCells < EMPTY_CELLS_MAX_COUNT.COLUMN) &&
    isEmpty
  );
};

const getBingoNumber = (usedNumbers, rowEmptyCells, columnEmptyCells) => {
  return isEmptyCell(rowEmptyCells, columnEmptyCells)
    ? null
    : getRandomInt(NUMBERS_RANGE.MIN, NUMBERS_RANGE.MAX, usedNumbers);
};

const increaseAdditionalCounters = (countersData, iteration) => ({
  ...countersData,
  [iteration]: !countersData[iteration] ? 1 : countersData[iteration] + 1
});

const generateBingoNumbers = () => {
  const rows = {};
  let cells = {};

  const usedNumbers = [];
  let rowsEmptyCellsMap = {};
  let columnsEmptyCellsMap = {};

  for (let i = 0; i < ROWS_MAX_COUNT; i++) {
    for (let j = 0; j < CELLS_ROW_MAX_COUNT; j++) {
      const idKey = uuid();
      const bingoNumber = getBingoNumber(
        usedNumbers,
        rowsEmptyCellsMap[i],
        columnsEmptyCellsMap[j]
      );

      if (typeof bingoNumber === 'number') {
        usedNumbers.push(bingoNumber);
      } else {
        rowsEmptyCellsMap = increaseAdditionalCounters(rowsEmptyCellsMap, i);
        columnsEmptyCellsMap = increaseAdditionalCounters(
          columnsEmptyCellsMap,
          j
        );
      }

      rows[i] = [...(rows[i] || []), idKey];
      cells = { ...cells, [idKey]: { bingoNumber, isChecked: false } };
    }
  }

  return { rows, cells };
};

const generateChips = () =>
  Array.from(new Array(20)).reduce(
    (acc, _, i) => ({
      ...acc,
      [i]: {
        id: i,
        isDragged: false,
        isDropped: false
      }
    }),
    {}
  );

export default {
  generateBingoNumbers,
  getRandomInt,
  generateChips
};
