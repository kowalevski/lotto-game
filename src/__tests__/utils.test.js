import { getRandomInt, getNamedNumbersWithLongName } from '../utils';

it('returns correct number', () => {
  const result = getRandomInt(0, 3, [2]);
  expect(typeof result === 'number').toBe(true);
  expect(result).toBeGreaterThanOrEqual(0);
  expect(result).toBeLessThanOrEqual(3);
  expect(result).not.toBe(2);
});

// useless test for useless function
it('returns named numbers with long names', () => {
  const namedNumbers = getNamedNumbersWithLongName();

  expect(namedNumbers).toMatchSnapshot();
});
