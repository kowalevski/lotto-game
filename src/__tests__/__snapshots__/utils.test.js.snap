// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns named numbers with long names 1`] = `
Array [
  Object {
    "name": "Three, Three",
    "number": 3,
  },
  Object {
    "name": "High Chairs",
    "number": 44,
  },
  Object {
    "name": "Half of hundred",
    "number": 50,
  },
  Object {
    "name": "Baker's dozen",
    "number": 13,
  },
  Object {
    "name": "For the first time",
    "number": 18,
  },
  Object {
    "name": "Back and forth",
    "number": 69,
  },
  Object {
    "name": "Ducklings, Geese Swans",
    "number": 22,
  },
  Object {
    "name": "Grandfather's neighbor",
    "number": 89,
  },
  Object {
    "name": "Grandfather",
    "number": 90,
  },
]
`;
